### ISSUE TITLE
Analysis plan

### ISSUE THREAD

**Issue by cgostic** [UBC-MDS/DSCI_522_group_415#2]
Starting a discussion thread for our analysis plan -- 

Initial thoughts on high-level plan:

- Use CountVectorizer to engineer features from each plate
    - We can do a gridsearch to optimize the length of ngrams used (i.e. 2,3 and 4 letter strings, 4,5 and 6 letter strings, etc)
- Run a multinomial NB model
- Either: Identify strongest predictors (which will be character strings) using predict_proba/feature weights, **OR** predict whether a new plate will be rejected

Suggestions?

**Comment by fkhan72** [comment]
Agree. The last part depends upon the type of research questions we want to answer.

**Comment by cgostic** [comment]
We also have to consider that we have an imbalanced dataset. There are 100,000+ examples of accepted plates, and ~2000 examples of rejected plates. I did a quick run of a MultinomialNB model on the vectorized data (default hyperparameters, that we'll need to tune) and saw a 98% accuracy, but basically, the model is always picking positive (the precision and recall for the rejected classification is very low). 

```
import pandas as pd 
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
import time

t1 = time.time()
df = pd.read_csv('data/vanity_plate_data.csv')
df = df.sample(frac=1).reset_index(drop=True).dropna()

X = df['plate']
y = df['outcome']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 1234)

cv = CountVectorizer(analyzer='char', ngram_range=(2,6))
X2 = cv.fit_transform(X_train)
X2_test = cv.transform(X_test)

mnb = MultinomialNB()
mnb.fit(X2, y_train)
print('trianing: ',mnb.score(X2, y_train))
print('testing: ',mnb.score(X2_test, y_test))
print(classification_report(y_train, mnb.predict(X2)))
print(time.time() - t1)
```

||precision|recall|f1|
|-|-|-|-|
|accepted|0.99|1.00|1.00|
|rejected|0.97|0.21|0.34|

I've read about a few simple methods for dealing with this:  
- Undersampling (randomly sampling the accepted plates to create a smaller dataset -- pulling 2000-3000 samples from the 100,000 available)
- Oversampling (replicating some points from the rejected set, but we'd have to replicate the rejected set 50 times to get a number close to the accepted set)

Any thoughts? There are definitely more complex (and likely accurate) ways to deal with imbalanced data, but for the scope of this project I'm of the opinion that either of the above is ok...
